File: src/main/java/examples/queens/Queens.java
package examples.queens;

import java.util.ArrayList;
import java.util.List;

import cspom.CSPOM;
import cspom.compiler.PredicateParseException;
import cspom.variable.CSPOMVariable;

public final class Queens {

	private static final int SIZE = 8;

	private Queens() {
	}

	public static CSPOM queens(final int nbQueens)
			throws PredicateParseException {

		final CSPOM problem = new CSPOM();

		final List<CSPOMVariable<Integer>> queens = new ArrayList<CSPOMVariable<Integer>>(
				nbQueens);

		for (int i = 0; i < nbQueens; i++) {
			queens.add(problem.interVar("q" + i, 0, nbQueens));
		}

		for (int i = nbQueens; --i >= 0;) {
			for (int j = i; --j >= 0;) {
				problem.ctr("neq(" + queens.get(i) + " , " + queens.get(j)
						+ ")");
				problem.ctr("neq(abs(sub(" + queens.get(i) + ","
						+ queens.get(j) + ")), " + Math.abs(i - j) + ")");
			}
		}

		return problem;
	}

	public static void main(final String[] args) throws PredicateParseException {
		System.out.println(queens(SIZE).toGML());
	}
}

File: src/main/scala/scalaexamples/queens/Queens.scala
package scalaexamples.queens;

import cspom.CSPOM
import CSPOM._
import cspom.variable.CSPOMVariable;
import cspom.compiler.ProblemCompiler

final object Queens {

  val SIZE = 8;

  def queens(nbQueens: Int) = CSPOM {

    val queens = (1 to nbQueens) map (i => interVar("Q" + i, 1, nbQueens)) toIndexedSeq

    for (Seq(i, j) <- (0 until nbQueens).combinations(2)) {
      ctr(queens(i) ≠ queens(j))
      ctr('abs(queens(i) - queens(j)) ≠ math.abs(i - j))
    }

  }

  def main(args: Array[String]) {
    val p = queens(SIZE)
    //ProblemCompiler.compile(p)
    println(p);
  }
}